Fixes #36715 - Speed up host fact retrieval when you only need a subset of facts

jeremylenz opened the pull request:
A user with about 3k hosts has a report that runs `host.facts` on each host multiple times. This was causing the report generation time to be over an hour, since `host.facts` runs a few `includes` which result in quite a few SQL queries:

```rb
fact_values.includes(:fact_name).collect do |fact|
        hash[fact.fact_name.name] = fact.value
      end
```

This PR adds a new argument to host.facts so you can pass specific fact names. It also changes `collect` to `each` in the above code because we don't use the resulting arrays, we just iterate and build the final hash.

The report template can look something like this:

```rb
<%- load_hosts(search: input('Hosts filter')).each_record do |host| -%>
<%- host_facts = host.facts(['net::interface::eth0::ipv4_address', 'network::hostname']) -%>
<%-   report_row(
        'Hostname': host_facts['network::hostname'],
        'IP': host_facts['net::interface::eth0::ipv4_address']
    ) -%>
<%- end -%>

<%= report_render -%>
```

jeremylenz committed:
Fixes #36715 - Speed up host fact retrieval

jeremylenz commented on app/models/host/base.rb:
```diff
@@ -228,6 +232,31 @@ def facts_hash
+      example '@host.facts_with_names([\'dmi::memory::size\', \'net::interface::eth0\']) # => { "dmi::memory::size"=>"16 GB", "net::interface::eth0"=>"00:50:56:8a:5c:3c" }', desc: 'Getting specific facts'
+    end
+    def facts_with_names(names_query = [], *args)
+      filtered_names = Set.new([names_query, *args].flatten.compact)
```
For some reason the report renderer splats the first argument, turning
```
host.facts_with_names(['fact1', 'fact2'])
```
into 
```
host.facts_with_names('fact1', 'fact2')
```
resulting in an ArgumentError if I don't handle it here.

parthaa commented:
I 'd recommend this approach instead of facts_with_names. In my benchmark tests it was way faster.

```diff
diff --git a/app/models/host/base.rb b/app/models/host/base.rb
index 36fbe72..4f1d980 100644
--- a/app/models/host/base.rb
+++ b/app/models/host/base.rb
@@ -221,13 +221,16 @@ module Host
       example '@host.facts["uptime"] # => "30 days"', desc: 'Getting specific fact value, +uptime+ in this case'
       aliases :facts
     end
-    def facts_hash
+    def facts_hash(fact_names = [])
       hash = {}
-      fact_values.includes(:fact_name).collect do |fact|
+      query = fact_values.includes(:fact_name)
+      query = query.where(fact_names: {name: fact_names}) if fact_names.present?
+      query.collect do |fact|
         hash[fact.fact_name.name] = fact.value
       end
       hash
     end

```

For example with my change
```ruby
fact_names = [
    'ansible_local::gls_ansible_ctd_posture::_ctd',
    'ansible_local::gls_ansible_timestemp::_timestemp',
    'ansible_local::gls_ansible_model::_model',
    'ansible_local::gls_ansible_mac::_mac_address',
    'ansible_local::gls_ansible_ipaddress::_gls_ansible_ipaddress',
    'ansible_local::gls_ansible_owner::_owner',
    'ansible_local::gls_ansible_lrt::_lab_lrt',
    'ansible_local::gls_ansible_hostname::_gls_ansible_hostname'
  ]

hashes = []
Benchmark.measure do 
  hashes = Host.all.map {|host| host.facts(fact_names)  }
end
=> #<Benchmark::Tms:0x000055fc793fe1b8 @label="", @real=4.964442664990202, @cstime=0.0, @cutime=0.0, @stime=0.068133, @utime=3.543512999999999, @total=3.611645999999999>
```
About 3.6 seconds for 2851 Hosts.

vs

```ruby

hashes1 = []
Benchmark.measure do 
  hashes1 = Host.limit(500).map {|host| host.facts_with_names(fact_names) }
end
=> #<Benchmark::Tms:0x000055fcf3c900b8 @label="", @real=39.416546951048076, @cstime=0.0, @cutime=0.0, @stime=0.7837529999999999, @utime=30.203799, @total=30.987552>
```

30 seconds for 500 host records

jeremylenz commented:
@parthaa updated :+1:

parthaa reviewed (approved):
This seems to work well for me. APJ

jeremylenz commented:
In our tests on a system with ~2900 hosts, this new approach has improved report generation time from over 1 hour to about 11-13 seconds.

![image](https://github.com/theforeman/foreman/assets/22042343/afc132ef-d2ad-4328-a6a4-02465c21816e)

evgeni commented on app/models/host/base.rb:
```diff
@@ -214,14 +214,18 @@ def set_comment(parser)
       hash = {}
-      fact_values.includes(:fact_name).collect do |fact|
+      query = fact_values.includes(:fact_name)
+      query = query.where(fact_names: {name: fact_names}) if filtered_names.present?
```
what's the difference between `fact_names` and `filtered_names`?

jeremylenz replied:
If I don't include the splatted args, I run into [this sort of thing](https://github.com/theforeman/foreman/pull/9819#discussion_r1320219359) when running the reports.

jeremylenz replied:
oh wait that `fact_names` should be `filtered_names`, good catch!

evgeni replied:
but you don't pass `filtered_names` to `where()`, you pass `fact_names` which in this case will be just the first thing?

jeremylenz replied:
(Read the previous two comments in reverse order.)

Updated

ekohl commented on app/models/host/base.rb:
```diff
@@ -214,14 +214,18 @@ def set_comment(parser)
-      fact_values.includes(:fact_name).collect do |fact|
+      query = fact_values.includes(:fact_name)
+      query = query.where(fact_names: {name: filtered_names}) if filtered_names.present?
+      query.each do |fact|
         hash[fact.fact_name.name] = fact.value
       end
       hash
```
Just a quick suggestion: I think this comes down to:
```suggestion
      query.to_h { |fact| [fact.fact_name.name, fact.value] }
```

But that also makes me think: do you need to retrieve all columns? I suspect you can use `pluck()`. Something like this probably works:
```suggestion
      query.pluck('fact_name.name', :value)
```

Then also drop `hash = {}`.

ekohl commented on app/models/host/base.rb:
```diff
@@ -214,14 +214,18 @@ def set_comment(parser)
       aliases :facts
     end
-    def facts_hash
+    def facts_hash(fact_names = [], *args)
```
What's `*args` here? Isn't the API effectively:
```suggestion
    def facts_hash(*fact_names)
```

And then drop `filtered_names = ...`

ekohl commented on app/models/host/base.rb:
```diff
@@ -220,15 +220,14 @@ def set_comment(parser)
-      hash
+    def facts_hash(*fact_names)
+      # keep it to one SQL query
+      query = if fact_names.present?
+                fact_values.includes(:fact_name).where(fact_names: {name: fact_names})
+              else
+                fact_values.includes(:fact_name)
+              end
```
Does it really trigger a query if you use:
```suggestion
      query = fact_values.includes(:fact_name)
      query = query.where(fact_names: {name: fact_names}) if fact_names.present?
```

I thought we used that query builder pattern quite a lot and it didn't trigger additional queries.

jeremylenz replied:
Yes, I confirmed it in the Rails console:

```
query = rhel9.fact_values.includes(:fact_name)
2023-09-18T15:45:40 [D|sql|]   FactValue Load (0.5ms)  SELECT "fact_values".* FROM "fact_values" WHERE "fact_values"."host_id" = $1  [["host_id", 3]]
2023-09-18T15:45:40 [D|sql|]   FactName Load (1.1ms)  SELECT "fact_names".* FROM "fact_names" WHERE "fact_names"."id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28, $29, $30, $31, $32, $33, $34, $35, $36, $37, $38, $39, $40, $41, $42, $43, $44, $45, $46, $47, $48, $49, $50, $51, $52, $53, $54, $55, $56, $57, $58, $59, $60, $61, $62, $63, $64, $65, $66, $67, $68, $69, $70, $71, $72, $73, $74, $75, $76, $77, $78, $79, $80, $81, $82, $83, $84, $85, $86, $87, $88, $89, $90, $91, $92, $93, $94, $95, $96, $97, $98, $99, $100, $101, $102, $103, $104, $105, $106, $107, $108, $109, $110, $111, $112, $113, $114, $115, $116, $117, $118, $119, $120, $121, $122, $123, $124, $125, $126, $127, $128, $129, $130, $131, $132, $133, $134, $135, $136, $137, $138, $139, $140, $141, $142, $143, $144, $145, $146, $147, $148, $149, $150, $151, $152, $153, $154, $155, $156, $157, $158, $159, $160, $161, $162, $163, $164, $165, $166, $167, $168, $169, $170, $171, $172, $173, $174) ORDER BY fact_names.name

query = fact_names.present? ? rhel9.fact_values.includes(:fact_name).where(fact_names: {name: fact_names}) : rhel9.fact_values.includes(:fact_name)
2023-09-18T15:47:33 [D|sql|]   SQL (0.8ms)  SELECT "fact_values"."id" AS t0_r0, "fact_values"."value" AS t0_r1, "fact_values"."fact_name_id" AS t0_r2, "fact_values"."host_id" AS t0_r3, "fact_values"."updated_at" AS t0_r4, "fact_values"."created_at" AS t0_r5, "fact_names"."id" AS t1_r0, "fact_names"."name" AS t1_r1, "fact_names"."updated_at" AS t1_r2, "fact_names"."created_at" AS t1_r3, "fact_names"."compose" AS t1_r4, "fact_names"."short_name" AS t1_r5, "fact_names"."type" AS t1_r6, "fact_names"."ancestry" AS t1_r7 FROM "fact_values" LEFT OUTER JOIN "fact_names" ON "fact_names"."id" = "fact_values"."fact_name_id" WHERE "fact_values"."host_id" = $1 AND "fact_names"."name" IN ($2, $3)
```

ekohl replied:
Now that I'm looking at it further: why do we even need `includes`? Now that we're using `pluck()` I think it's redundant.

jeremylenz replied:
Seems it's needed:

![image](https://github.com/theforeman/foreman/assets/22042343/b1c89bd7-813a-4efe-9fa7-28edc0492304)

ekohl replied:
Isn't `query.joins(:fact_name)` the right way to go here?

jeremylenz replied:
That seems to work :+1: 

Updated

ekohl commented on app/models/host/base.rb:
```diff
@@ -214,17 +214,20 @@ def set_comment(parser)
       desc 'Note that available facts depend on what facts have been uploaded to Foreman,
            typical sources are Puppet facter, subscription manager etc.
            The facts can be out of date, this macro only provides access to the value stored in the database.'
+      optional :fact_names, Array, desc: 'A list of fact names to return. If empty all facts are returned'
```
@ofedoren is this also correct documentation for `*args` parameters, or is there a specific tag for that?

ofedoren replied:
Agh, my bad for not noticing this earlier :/

Well, it's not as wrong. It's still quite optional :) But for `*args` params, I'd use one of `list`/`splat`/`rest` instead of `optional`.

ekohl replied:
Could you submit a follow up PR?

ofedoren replied:
Sure: https://github.com/theforeman/foreman/pull/9850

ekohl reviewed (dismissed):
It'd be interested to see some benchmarks of the new code in the latest iteration.

jeremylenz commented:
I ran it again and report generation time (on our system with ~2900 hosts) is down to 9 seconds, not bad!
![image](https://github.com/theforeman/foreman/assets/22042343/4e0763ee-a40d-4b3c-baad-c33d60b940b0)

jeremylenz commented:
Squashed

ekohl commented:
[test katello]

ekohl commented on app/models/host/base.rb:
```diff
@@ -214,17 +214,20 @@ def set_comment(parser)
-      hash
+    def facts_hash(*fact_names)
+      # keep it to one SQL query
+      query = if fact_names.present?
```
I'm much more used to plain Ruby where `Array.present?` doesn't exist. Would it make sense to stick to plain Ruby and use `.any?`:
```suggestion
      query = if fact_names.any?
```

ekohl reviewed (commented):
I like the improvement (looks like join & pluck shaved off another few seconds), but I wonder if there is test coverage for this new functionality. Can we also assert that it only performs a single DB query?

jeremylenz commented:
> test coverage for this new functionality. Can we also assert that it only performs a single DB query?

Did a bit of research and it looks like there is a "bullet" gem that helps with this, but I didn't want to add any new gems. So I added a helper method in test_helper; see what you think.

Turns out `.pluck` was adding an additional SQL query being on a separate line, so I fixed that as well.

I also improved `facts` to cache the results for a single host. This is because a lot of reports call `host.facts[xxx]` repeatedly, and not all users may immediately switch to the new `facts(['fact_name1', 'fact_name2'])` syntax:

```
'CTD': host.facts['ansible_local::gls_ansible_ctd_posture::_ctd'],
 'TimeStemp': host.facts['ansible_local::gls_ansible_timestemp::_timestemp'],
 'Model': host.facts['ansible_local::gls_ansible_model::_model'],
```

Added tests to cover all the new functionality.

evgeni commented on test/test_helper.rb:
```diff
@@ -119,6 +119,16 @@ def after_teardown
+  def assert_sql_queries(num_of_queries)
+    queries = []
+    ActiveSupport::Notifications.subscribe("sql.active_record") do |name, start, finish, id, payload|
+      queries << payload[:sql] if payload[:sql] =~ /SELECT/
```
shouldn't this also catch `UPDATE` and `DROP` and `INSERT` statements?

(but probably out of scope right now)

jeremylenz replied:
It could be argued that those are SQL "statements" and not "queries" :)

evgeni replied:
Given `SELECT` can change data (hello triggers!), I see no difference. But as said, this is probably more than was asked in the original comment by Ewoud :)

evgeni commented:
I had promised you a review when I am back, but it seems Ewoud did beat me, so we're good :)

parthaa commented on test/test_helper.rb:
```diff
@@ -119,6 +119,16 @@ def after_teardown
     @clear_plugins = nil
   end
+
+  def assert_sql_queries(num_of_queries)
```
```suggestion
  def assert_sql_queries(num_of_queries, match = /SELECT/)
```

followed by
```ruby
 queries << payload[:sql] if payload[:sql] =~ match
```

jeremylenz replied:
Good idea. Updated :+1:

jeremylenz commented:
Seems this will need another approving review since I added those tests.

jeremylenz commented:
@evgeni @ekohl any more comments?
